constantine/commitments/ipa/barycentric_form.nim

```diff
@@ -0,0 +1,187 @@
+# Constantine
+# Copyright (c) 2018-2019    Status Research & Development GmbH
+# Copyright (c) 2020-Present Mamy André-Ratsimbazafy
+# Licensed and distributed under either of
+#   * MIT license (license terms in the root directory or at http://opensource.org/licenses/MIT).
+#   * Apache v2 license (license terms in the root directory or at http://www.apache.org/licenses/LICENSE-2.0).
+# at your option. This file may not be copied, modified, or distributed except according to those terms.
+import 
+ math,
+ ./helper_types,
+ ../../../constantine/math/config/[type_ff, curves],
+ ../../../constantine/math/elliptic/[ec_twistededwards_projective],
+ ../../../constantine/math/arithmetic/[finite_fields],
+ ../../../constantine/math/arithmetic
+
+# ############################################################
+#
+#       Barycentric Form using Precompute Optimisation
+#
+# ############################################################
+
+# Please refer to https://hackmd.io/mJeCRcawTRqr9BooVpHv5g 
+
+func newPrecomputedWeights* [PrecomputedWeights] (res: var PrecomputedWeights)=
+ 
+ var midpoint: uint64 = 256
+ for i in uint64(0)..<midpoint:
+  var weights {.noInit.}: EC_P_Fr
+  weights.computeBarycentricWeights(i) 
+
+  var inverseWeights {.noInit.}: EC_P_Fr
+  inverseWeights.inv(weights)
+
+  res.barycentricWeights[i] = weights
+  res.barycentricWeights[i+midpoint] = inverseWeights
+
+  midpoint = uint64(DOMAIN) - 1
+
+  for i in 1..<DOMAIN:
+   var k {.noInit.}: EC_P_Fr
+   var i_bg {.noInit.} : matchingOrderBigInt(Banderwagon)
+   i_bg.setUint(uint64(i))
+   k.fromBig(i_bg)
+
+   k.inv(k)
+
+
+   var neg_k : EC_P_Fr
+   var zero : EC_P_Fr
+   zero.setZero()
+   neg_k.diff(zero, k)
+   res.invertedDomain[i-1] = k
+   res.invertedDomain[(i-1) + int(midpoint)] = neg_k
+
+
+func computeBarycentricWeights* [EC_P_Fr] (res: var EC_P_Fr, element : uint64) = 
+ if element <= uint64(DOMAIN):
+
+  var domain_element_Fr: EC_P_Fr
+  var bigndom : matchingOrderBigInt(Banderwagon)
+  bigndom.setUint(uint64(element))
+  domain_element_Fr.fromBig(bigndom)
+
+  res.setOne()
+
+  for i in uint64(0)..<uint64(DOMAIN):
+    if i==element:
```

Comment:
The idea, as far as I understand, barycentric coefficients are used to cache the points that are in `InDomain` as per spec, here most of the times whenever challenge scalar generates points (later to be evaluated), they are _mostly_ out of domain points because the Probability (in domain points) = 1-256/2^253 (which is 253 bit scalar field), which is very low, and out of domain is 1 - that low probability value. So for now I haven't specified anything as most points that needs to be evaluated will be mostly out of domain. Nevertheless, I shall prefix `inDomain` with Barycentric